### PR TITLE
Add array_remove Presto scalar function

### PR DESCRIPTION
**Pull request by myandpr**
# Description
Presto support [array_remove](https://prestodb.io/docs/current/functions/array.html#:~:text=0%20is%20returned.-,array_remove,-(x%2C ) 
and 
Spark support [array_remove](https://spark.apache.org/docs/latest/api/sql/#array_remove) since 2.4.0</br>
let's add it to Velox.</br>
### array_remove(x, element) → array
* Not Implemented for Presto and Spark.
* Remove all elements that equal element from array x.
### For example:
```
> SELECT array_remove(array(1, 2, 3, null, 3), 3);
 [1,2,null]
```

### CODE REVIEW

**Comment by myandpr** [comment]
Hi @zacw7, would you help review this PR if you have time ?

**Comment by myandpr** [comment]
Hi @mbasmanova , would you help review this PR if you hava time? This Pr is about Adding array function for presto.

File: velox/functions/prestosql/ArrayFunctions.h [hunk 1]
```diff
@@ -639,4 +639,38 @@ struct ArrayFlattenFunction {
...
+/// This class implements the array remove function.
+///
+/// DEFINITION:
+/// array_remove(x, element) → array
+/// Remove all elements that equal element from array x.
+template <typename TExecCtx, typename T>
+struct ArrayRemoveFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(TExecCtx);
+
+  FOLLY_ALWAYS_INLINE void call(
```
Comment (laithsakka) [inline velox/functions/prestosql/ArrayFunctions.h:652]: no need for FOLLY_ALWAYS_INLINE since inline does not have much value here

File: velox/functions/prestosql/ArrayFunctions.h [hunk 2]
```diff
@@ -639,4 +639,38 @@ struct ArrayFlattenFunction {
...
+      out_type<Array<T>>& out,
+      const arg_type<Array<T>>& array,
```
Comment (laithsakka) [inline velox/functions/prestosql/ArrayFunctions.h:654]: lets also add the generic (non-fast path ) support for all types. 
for examples see ArrayTrim function. 

https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/ArrayFunctions.h
you can check the equal function to compare two items or direcly use compare 
(look at eq function here 
https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/Comparisons.h

File: velox/functions/prestosql/ArrayFunctions.h [hunk 3]
```diff
@@ -639,4 +639,38 @@ struct ArrayFlattenFunction {
...
+      if (element != item.value()) {
+        elementCount++;
+      }
+    }
+    out.reserve(elementCount);
+
+    for (const auto& item : array) {
+      if (element != item.value()) {
+        if (item.has_value()) {
+          auto& newItem = out.add_item();
```
Comment (laithsakka) [inline velox/functions/prestosql/ArrayFunctions.h:667]: FYI: for primitives you can directly do out.push_back(item.value()) and for nulls you can do out.push_back(std::nullopt).

it would be nice to allow out.add_item(item); to be handled where item is optional accessor but we do have that 
now,. will file task for it. that will reduce 6 lines here to one.

File: velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp [hunk 4]
```diff
@@ -198,6 +207,13 @@ void registerArrayFunctions(const std::string& prefix) {
   registerArrayHasDuplicatesFunctions<int128_t>(prefix);
   registerArrayHasDuplicatesFunctions<Varchar>(prefix);
 
+  registerArrayRemoveFunctions<int8_t>(prefix);
+  registerArrayRemoveFunctions<int16_t>(prefix);
+  registerArrayRemoveFunctions<int32_t>(prefix);
+  registerArrayRemoveFunctions<int64_t>(prefix);
+  registerArrayRemoveFunctions<int128_t>(prefix);
+  registerArrayRemoveFunctions<Varchar>(prefix);
```
Comment (laithsakka) [inline velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp:215]: for Varchar you can optimize it by avoid copying strings see for example ArrayTrim function.

**Review by laithsakka — CHANGES_REQUESTED** [review]

File: velox/docs/functions/presto/array.rst [hunk 5]
```diff
@@ -127,6 +127,13 @@ Array Functions
 
     If ``instance > 0``, returns the position of the ``instance``-th occurrence of the ``element`` in array ``x``. If ``instance < 0``, returns the position of the ``instance``-to-last occurrence of the ``element`` in array ``x``. If no matching element instance is found, 0 is returned.
 
+.. function:: array_remove(x, element) -> array
+
+    Remove all elements that equal element from array x.
```
Comment (zacw7) [inline velox/docs/functions/presto/array.rst:132]:
```suggestion
    Remove all elements that equal ``element`` from array ``x``.
```

File: velox/functions/prestosql/tests/ArrayRemoveTest.cpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+#include "velox/functions/prestosql/tests/utils/FunctionBaseTest.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::test;
+using namespace facebook::velox::exec;
+using namespace facebook::velox::functions::test;
+
+namespace {
+
+class ArrayRemoveTest : public FunctionBaseTest {
```
Comment (zacw7) [inline velox/functions/prestosql/tests/ArrayRemoveTest.cpp:26]: Can we add some tests to cover the following cases:
```SQL
SELECT ARRAY_REMOVE(ARRAY[1, 1, NULL, 2, 3], NULL);
SELECT ARRAY_REMOVE(ARRAY[ARRAY[1, 1], NULL, ARRAY[2, 2]], ARRAY[1, 2]);
```

File: velox/functions/prestosql/ArrayFunctions.h (hunk 1 above)
Comment (myandpr) [reply velox/functions/prestosql/ArrayFunctions.h]: @laithsakka remove FOLLY_ALWAYS_INLINE

Done.

File: velox/functions/prestosql/ArrayFunctions.h (hunk 2 above)
Comment (myandpr) [reply velox/functions/prestosql/ArrayFunctions.h]: added generic support for all types.

Done.

File: velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp (hunk 4 above)
Comment (myandpr) [reply velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp]: @laithsakka , thanks for you comments.
optimized Varchar function by setNoCopy()

Done..

File: velox/docs/functions/presto/array.rst (hunk 5 above)
Comment (myandpr) [reply velox/docs/functions/presto/array.rst]: Thanks for your comments.

Done.

File: velox/functions/prestosql/tests/ArrayRemoveTest.cpp (hunk 6 above)
Comment (myandpr) [reply velox/functions/prestosql/tests/ArrayRemoveTest.cpp:26]: Thank your comments.
that's a good idea, I'll add these two ut later

Comment (myandpr) [reply velox/functions/prestosql/tests/ArrayRemoveTest.cpp:26]: @zacw7 I have added these two tests. Thanks for your suggests.

**Comment by zacw7** [comment]
Could you please squash the commits?

**Comment by myandpr** [comment]
> Could you please squash the commits?

Ok, I have squashed these commits.

File: velox/functions/prestosql/ArrayFunctions.h (hunk 3 above)
Comment (zacw7) [reply velox/functions/prestosql/ArrayFunctions.h:679]: Has this comment been addressed?

Comment (myandpr) [reply velox/functions/prestosql/ArrayFunctions.h:679]: Done, Currently I used `call(Out& out, const In& inputArray,...`, so still use  `out.add_item()`(in this case, out.push_back() doesn`t work)

**Comment by myandpr** [comment]
Hi @laithsakka , I have resolved all the comments. Would you help take a look if there is anything that needs to be changed?many thanks! @

**Comment by myandpr** [comment]
Hi @zacw7 , could we merge this PR? is any problem I need to solve? thanks!

File: velox/functions/prestosql/ArrayFunctions.h [hunk 7]
```diff
@@ -17,6 +17,7 @@
 
 #include "velox/functions/Udf.h"
 #include "velox/functions/prestosql/CheckedArithmetic.h"
+#include "velox/functions/sparksql/Comparisons.h"
```
Comment (zacw7) [inline velox/functions/prestosql/ArrayFunctions.h:20]: Why is this needed?

Comment (myandpr) [reply velox/functions/prestosql/ArrayFunctions.h]: sorry, it`s redundant.

Deleted.

**Comment by zacw7** [comment]
Please squash the commits. Thanks.

**Comment by myandpr** [comment]
> Please squash the commits. Thanks.

Squashed.

**Comment by zacw7** [comment]
@laithsakka do you mind taking a look at the fuzzer failure?

**Comment by mbasmanova** [comment]
The Fuzzer fails on expression that uses array_remove. It is this PR's author's responsibility to investigate and fix.

```
I20230713 20:04:03.235684 229671 ExpressionFuzzer.cpp:1312] ==============================> Started iteration 55 (seed: 4090465574)
I20230713 20:04:03.236156 229671 ExpressionVerifier.cpp:81] Executing expression 0 : array_min(array_sort(array_remove("c0",is_infinite("c1"))))
I20230713 20:04:03.236178 229671 ExpressionVerifier.cpp:30] 2 vectors as input:
```

**Comment by myandpr** [comment]
@mbasmanova ,I have few confusion since I don't know much about the mechanics of Fuzzer.
1.  Why didn't this error occur in the previous Fuzzer test? Every input is same?
2. From the following logs, I can't see the specific input and error output for array_remove().So is there any way to clarify the error more clearly to better debug?
```
20230713 20:04:03.235602 229671 ExpressionVerifier.cpp:69] All results match.
I20230713 20:04:03.235641 229671 ExpressionFuzzer.cpp:1357] ==============================> Done with iteration 54
I20230713 20:04:03.235684 229671 ExpressionFuzzer.cpp:1312] ==============================> Started iteration 55 (seed: 4090465574)
I20230713 20:04:03.236156 229671 ExpressionVerifier.cpp:81] Executing expression 0 : array_min(array_sort(array_remove("c0",is_infinite("c1"))))
I20230713 20:04:03.236178 229671 ExpressionVerifier.cpp:30] 2 vectors as input:
I20230713 20:04:03.236186 229671 ExpressionVerifier.cpp:32] 	[ARRAY ARRAY<BOOLEAN>: 100 elements, no nulls]
I20230713 20:04:03.236197 229671 ExpressionVerifier.cpp:32] 	[DICTIONARY DOUBLE: 100 elements, 14 nulls], [DICTIONARY DOUBLE: 100 elements, 12 nulls], [DICTIONARY DOUBLE: 100 elements, 11 nulls], [DICTIONARY DOUBLE: 100 elements, 10 nulls], [FLAT DOUBLE: 100 elements, 10 nulls]
E20230713 20:04:03.238248 229671 Exceptions.h:68] Line: ../../velox/expression/tests/ExpressionVerifier.cpp:61, Function:operator(), Expression: left->equalValueAt(right.get(), row, row) Different results at idx '1': 'null' vs. 'true', Source: RUNTIME, ErrorCode: INVALID_STATE
I20230713 20:04:03.239174 229671 ExpressionVerifier.cpp:327] Persisted input: --fuzzer_repro_path /tmp/fuzzer_repro/velox_expressionVerifier_zTfdOH --input_path /tmp/fuzzer_repro/velox_expressionVerifier_zTfdOH/input_vector --result_path /tmp/fuzzer_repro/velox_expressionVerifier_zTfdOH/result_vector --sql_path /tmp/fuzzer_repro/velox_expressionVerifier_zTfdOH/sql
terminate called after throwing an instance of 'facebook::velox::VeloxRuntimeError'
  what():  Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Different results at idx '1': 'null' vs. 'true'
Retriable: False
Expression: left->equalValueAt(right.get(), row, row)
Function: operator()
File: ../../velox/expression/tests/ExpressionVerifier.cpp
Line: 61
Stack trace:
# 0  _ZN8facebook5velox7process10StackTraceC1Ei
# 1  _ZN8facebook5velox14VeloxException5State4makeIZNS1_C4EPKcmS5_St17basic_string_viewIcSt11char_traitsIcEES9_S9_S9_bNS1_4TypeES9_EUlRT_E_EESt10shared_ptrIKS2_ESA_SB_
# 2  _ZN8facebook5velox14VeloxExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 3  _ZN8facebook5velox17VeloxRuntimeErrorC2EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bS7_
# 4  _ZN8facebook5velox6detail14veloxCheckFailINS0_17VeloxRuntimeErrorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRKNS1_18VeloxCheckFailArgsET0_
# 5  _ZZN8facebook5velox4test12_GLOBAL__N_114compareVectorsERKSt10shared_ptrINS0_10BaseVectorEES7_RKNS0_17SelectivityVectorEENKUliE0_clEi
# 6  _ZNK8facebook5velox17SelectivityVector15applyToSelectedIZNS0_4test12_GLOBAL__N_114compareVectorsERKSt10shared_ptrINS0_10BaseVectorEES9_RKS1_EUliE0_EEvT_
# 7  _ZN8facebook5velox4test12_GLOBAL__N_114compareVectorsERKSt10shared_ptrINS0_10BaseVectorEES7_RKNS0_17SelectivityVectorE
# 8  _ZN8facebook5velox4test18ExpressionVerifier6verifyERKSt6vectorISt10shared_ptrIKNS0_4core10ITypedExprEESaIS8_EERKS4_INS0_9RowVectorEEOS4_INS0_10BaseVectorEEbS3_IiSaIiEE
# 9  _ZN8facebook5velox4test16ExpressionFuzzer2goEv
# 10 _ZN8facebook5velox4test16expressionFuzzerESt13unordered_mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt6vectorIPKNS0_4exec17FunctionSignatureESaISD_EESt4hashIS8_ESt8equal_toIS8_ESaISt4pairIKS8_SF_EEEm
# 11 _ZN12FuzzerRunner3runERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEmRKSt13unordered_setIS5_St4hashIS5_ESt8equal_toIS5_ESaIS5_EES7_
# 12 main
# 13 __libc_start_main
# 14 _start

*** Aborted at 1689278643 (Unix time, try 'date -d @1689278643') ***
*** Signal 6 (SIGABRT) (0x38127) received by PID 229671 (pthread TID 0x7f52541ff2c0) (linux TID 229671) (maybe from PID 229671, UID 0) (code: -6), stack trace: ***
(error retrieving stack trace)
/bin/bash: line 16: 229671 Aborted                 (core dumped) _build/debug/velox/expression/tests/velox_expression_fuzzer_test --seed ${RANDOM} --enable_variadic_signatures --velox_fuzzer_enable_complex_types --lazy_vector_generation_ratio 0.2 --velox_fuzzer_enable_column_reuse --velox_fuzzer_enable_expression_reuse --max_expression_trees_per_step 2 --retry_with_try --enable_dereference --duration_sec 60 --logtostderr=1 --minloglevel=0 --repro_persist_path=/tmp/fuzzer_repro

Exited with code exit status 134
```

**Comment by mbasmanova** [comment]
@myandpr Maybe start by reading the Fuzzer documentation: https://facebookincubator.github.io/velox/develop/testing/fuzzer.html

Let me know if you have questions after I have read that.

**Comment by myandpr** [comment]
> @myandpr Maybe start by reading the Fuzzer documentation: https://facebookincubator.github.io/velox/develop/testing/fuzzer.html
> 
> 
> 
> Let me know if you have questions afterwards.

Thanks.

File: velox/functions/prestosql/ArrayFunctions.h (hunk 3 above)
Comment (laithsakka) [reply velox/functions/prestosql/ArrayFunctions.h:678]: out.push_back(item.value()) did not work?

File: velox/functions/prestosql/ArrayFunctions.h [hunk 8]
```diff
@@ -639,4 +639,74 @@ struct ArrayFlattenFunction {
...
+          newItem = item.value();
+        } else {
+          out.add_null();
+        }
+      }
+    }
+  }
+
+  // Generic implementation.
+  FOLLY_ALWAYS_INLINE void call(
```
Comment (laithsakka) [inline velox/functions/prestosql/ArrayFunctions.h:668]: no need for folly inline here

File: velox/functions/prestosql/ArrayFunctions.h [hunk 9]
```diff
@@ -639,4 +639,74 @@ struct ArrayFlattenFunction {
...
+/// Remove all elements that equal element from array x.
+template <typename T>
+struct ArrayRemoveFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T)
+
+  // Fast path for primitives.
+  template <typename Out, typename In, typename E>
+  void call(Out& out, const In& inputArray, E element) {
+    for (const auto& item : inputArray) {
+      if (element != item.value()) {
```
Comment (laithsakka) [inline velox/functions/prestosql/ArrayFunctions.h:656]: what if item is null?
it.value() is not safe to always be called, you should first check item.has_value();

File: velox/functions/prestosql/ArrayFunctions.h [hunk 10]
```diff
@@ -639,4 +639,74 @@ struct ArrayFlattenFunction {
...
+
+  // Generic implementation.
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Array<Generic<T1>>>& out,
+      const arg_type<Array<Generic<T1>>>& inputArray,
+      const arg_type<Generic<T1>>& element) {
+    static constexpr CompareFlags kFlags = {
+        false, false, /*euqalsOnly*/ true, true /*stopAtNull*/};
+    for (const auto& item : inputArray) {
+      auto result = element.compare(item.value(), kFlags);
```
Comment (laithsakka) [inline velox/functions/prestosql/ArrayFunctions.h:675]: same item.has_value() can be false.

**Comment by laithsakka** [comment]
what should happen when element is Null? in your implementation the whole result is null but I assume you want 
to compare it with every element. 
In that case you probably want to use callNullable, so default null is not triggered

**Comment by laithsakka** [comment]
@myandpr
are you on slack?

**Comment by myandpr** [comment]
> @myandpr
> 
> are you on slack?

No,i tried to send an email to velox@meta.com to get access to Velox Slack Channel following the document,but I didn't receive a response.

**Comment by myandpr** [comment]
> check https://facebookincubator.github.io/velox/develop/scalar-functions.html#null-behavior

Thanks, i will read the link

**Comment by zacw7** [comment]
> > check https://facebookincubator.github.io/velox/develop/scalar-functions.html#null-behavior
> 
> Thanks, i will read the link

Hey @myandpr , wondering what's the current status on the fuzzer test fix? Do you need any help?

**Comment by zacw7** [comment]
@myandpr Closing this PR in 3 days.  I'll take over the array_remove implementation to unblock us. Thanks.

**Comment by myandpr** [comment]
>

**Comment by myandpr** [comment]
> @myandpr Closing this PR in 3 days. I'll take over the array_remove implementation to unblock us. Thanks.

@zacw7 Sorry for taking so long to reply. I am still in the hospital right now due to some accident.
I may not be able to complete this PR.
It would be great if you could take over the 'array_remove' implementation.
thank you very much!

**Comment by myandpr** [comment]
> @myandpr Closing this PR in 3 days. I'll take over the array_remove implementation to unblock us. Thanks.

@zacw7 Sorry for taking so long to reply. I am still in the hospital right now due to some accident.
I may not be able to complete this PR.
It would be great if you could take over the 'array_remove' implementation.
many thanks Zac!
